File: frontend/src/types/analysis.ts
export interface ColumnType {
  name: string
  type: string
}

export interface NumericStatistics {
  name: string
  count: number
  unique: number
  mean: number | null
  median: number | null
  std: number | null
  min: number | null
  q1: number | null
  q3: number | null
  max: number | null
}

export interface AnalysisResult {
  filename: string
  rows: number
  columns: number
  column_names: string[]
  column_types: ColumnType[]
  numeric_statistics: NumericStatistics[]
  numeric_histograms: NumericHistogram[]
  missing_values: Record<string, number>
  missing_percentage: number
  duplicates: number
  size_bytes: number
}

export interface AnalysisHistoryItem {
  id: number
  filename: string
  rows: number
  columns: string[]
  created_at: string
}

export interface NumericHistogram {
  name: string
  labels: string[]
  counts: number[]
}

File: backend/app/schemas/analysis.py
from datetime import datetime

from pydantic import BaseModel, ConfigDict


class ColumnTypeResponse(BaseModel):
    name: str
    type: str


class NumericStatisticsResponse(BaseModel):
    name: str
    count: int
    unique: int
    mean: float | None
    median: float | None
    std: float | None
    min: float | None
    q1: float | None
    q3: float | None
    max: float | None


class NumericHistogramResponse(BaseModel):
    name: str
    labels: list[str]
    counts: list[int]


class AnalysisResponse(BaseModel):
    filename: str
    rows: int
    columns: int
    column_names: list[str]
    column_types: list[ColumnTypeResponse]
    numeric_statistics: list[NumericStatisticsResponse]
    numeric_histograms: list[NumericHistogramResponse]
    missing_values: dict[str, int]
    missing_percentage: float
    duplicates: int
    size_bytes: int


class AnalysisHistoryResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: int
    filename: str
    rows: int
    columns: list[str]
    created_at: datetime